contracts/VestingEscrow.vy

```diff
@@ -0,0 +1,371 @@
+    self.start_time = start_time
+    self.end_time = end_time
+    self.cliff_length = cliff_length
+
+    assert self.token.transferFrom(
+        msg.sender, self, amount, default_return_value=True
+    ), "funding failed"
+
+    self.total_locked = amount
+    self.recipient = recipient
+    self.disabled_at = end_time  # Set to maximum time
+    self.factory = factory
+    log VestingEscrowInitialized(
+        factory,
+        recipient,
+        token,
+        amount,
+        start_time,
+        end_time,
+        cliff_length,
+        is_fully_revokable,
+    )
+
+    return True
+
+
+@internal
+@view
+def _total_vested_at(time: uint256) -> uint256:
+    start: uint256 = self.start_time
+    end: uint256 = self.end_time
+    locked: uint256 = self.total_locked
+    if time < start + self.cliff_length:
+        return 0
+    return min(locked * (time - start) / (end - start), locked)
+
+
+@internal
+@view
+def _unclaimed(time: uint256) -> uint256:
+    if self.is_fully_revoked:
+        return 0
+    return self._total_vested_at(time) - self.total_claimed
+
+
+@external
+@view
+def unclaimed() -> uint256:
+    """
+    @notice Get the number of unclaimed, vested tokens for recipient
+    """
+    # NOTE: if `revoke_unvested` is activated, limit by the activation timestamp
+    return self._unclaimed(min(block.timestamp, self.disabled_at))
+
+
+@internal
+@view
+def _locked(time: uint256 = block.timestamp) -> uint256:
+    if time >= self.disabled_at:
+        return 0
+    return self.total_locked - self._total_vested_at(time)
+
+
+@external
+@view
+def locked() -> uint256:
+    """
+    @notice Get the number of locked tokens for recipient
+    """
+    return self._locked(min(block.timestamp, self.disabled_at))
+
+
+@external
+def claim(
+    beneficiary: address = msg.sender, amount: uint256 = max_value(uint256)
+):
+    """
+    @notice Claim tokens which have vested
+    @param beneficiary Address to transfer claimed tokens to
+    @param amount Amount of tokens to claim
+    """
+    self._check_sender_is_recipient()
+
+    claim_period_end: uint256 = min(block.timestamp, self.disabled_at)
+    claimable: uint256 = min(self._unclaimed(claim_period_end), amount)
+    self.total_claimed += claimable
+
+    assert self.token.transfer(
+        beneficiary, claimable, default_return_value=True
+    ), "transfer failed"
+
+    log Claim(self.recipient, beneficiary, claimable)
+
+
+@external
+def revoke_unvested():
+    """
+    @notice Disable further flow of tokens and revoke the unvested part to owner
+    """
+    self._check_sender_is_owner_or_manager()
+    # NOTE: Revoking more than once is futile
+
+    revokable: uint256 = self._locked()
+    self.disabled_at = block.timestamp
+
+    assert self.token.transfer(
+        self._owner(), revokable, default_return_value=True
+    ), "transfer failed"
+
+    log UnvestedTokensRevoked(msg.sender, self.recipient, revokable)
+
+
+@external
+def revoke_all():
```

Comment:
maybe it makes sense to call some kind of `snapshot_clear_delegate` here?

and perhaps in the case of a complete claim?

Reply:
Snapshot relies on the current balance. If tokens are revoked or claimed than they are no longer available for the voting